DM-34855: Add vectorized way of getting URIs for many datasets

n8pease opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

n8pease committed:
Allow ignoring ID column in astropy table comparison

This was added to other tests by Andy S. in
ebb1cc63f5179673c940db636f59f6adfda01c51

I don't know why it wasn't added to this test case, but it's
fails now when running the test alone (but not when running scons)
because the dataset id gets changed to a uuid style value during
test execution.

n8pease committed:
add release note

timj commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -707,6 +724,34 @@ def transfer_from(
         raise NotImplementedError(f"Datastore {type(self)} must implement a transfer_from method.")
 
+    def getManyURIs(
+        self, refs: Iterable[DatasetRef], predict: bool = False, throwIfNotFound: bool = True
```
I'm trying to switch to things like `throw_if_not_found` instead. How about a more positive spin: `allow_missing`, defaulting to `False`?

I initially didn't understand why we needed it but I see now that it's critical for ChainedDatastore to be able to function.

timj commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -707,6 +724,34 @@ def transfer_from(
+        URIs : `dict` of [`DatasetRef`, `DatasetRefUris`]
+            A dict of primary and component URIs, indexed by the passed-in
+            refs.
+        """
```
Add a `Raises` section.

timj commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -194,6 +194,23 @@ def commit(self) -> None:
 
 
+@dataclasses.dataclass
+class DatasetRefURIs:
```
I'm not sure how easy it would be for us to change `getURIs` at this point, unless we had a way for this dataclass to take 0 and 1 index or unpack as a tuple for existing callers.

timj commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -707,6 +724,34 @@ def transfer_from(
+            References to the required datasets.
+        predict : `bool`, optional
+            If the datastore does not know about a dataset, should it
+            return a predicted URI or not?
```
Missing documentation of the boolean parameter. Presumably it's only relevant if `predict=False`?

timj commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -707,6 +724,34 @@ def transfer_from(
+                uris[ref] = DatasetRefURIs(*self.getURIs(ref, predict=predict))
+            except FileNotFoundError:
+                if throwIfNotFound:
+                    raise
```
I think I would prefer it if the code accumulated the missing refs and then reported them all at the end rather than only reporting the first one. In theory the list could get large but we should be able to say something like:

> Missing 5 DatasetRefs from datastore out of 400 and predict=False

timj commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -501,6 +502,34 @@ def _finishIngest(
+                missingRefs -= gotUris.keys()
+                uris.update(gotUris)
+                if not missingRefs:
+                    break
```
This breaks out the first loop but it will still cause the outer loop to keep running so the first datastore will be called again with no refs.

timj commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -501,6 +502,34 @@ def _finishIngest(
+        # datastore if it actually exists in a later datastore, so in that
+        # case check all datastores with predict=False first, and then try
+        # again with predict=True.
+        for p in (False, True) if predict else (False,):
```
I've not seen this style before. Interesting.

timj commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -501,6 +502,34 @@ def _finishIngest(
+                    break
+
+        if missingRefs and throwIfNotFound:
+            raise FileNotFoundError("Dataset(s) {} not in this datastore.".format(missingRefs))
```
Should test the missing refs branches.

timj commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1762,6 +1763,105 @@ def getURI(self, ref: DatasetRef, predict: bool = False) -> ResourcePath:
+
+        records = self._get_stored_records_associated_with_refs(refs)
+
+        missingRefs = [ref for ref in refs if ref.id not in records.keys()]
```
Is there an overhead to calling `records.keys()` in a loop? I'm assuming it's some kind of internal state that has low overhead each call so doesn't need to be stored in a variable first.

n8pease replied:
`records` is just a `dict` so the result of `records.keys()` is a `dict_keys`, which is designed to be iterated over and behaves something like a generator if I understand it correctly.

For that matter if this was going to be unchanged I probably should have changed the whole statement from a list comprehension to a generator IE
```
missingRefs = (ref for ref in refs if ref.id not in records.keys())
```
because there's no need to materialize the whole `missingRefs` list (I did think it was helpful to name the list though, instead of using it directly in line 1782).

But, as you note on line 1818 maybe we shouldn't iterate over the whole list of refs again, so this will want some rewriting to keep track of the not-missing refs.

timj replied:
I think `records.keys()` is set-like so `in` should be fast. I was mainly wondering whether there is a cost to python to create it or if it's a read-only view of something that is free to return each time.

n8pease replied:
oh, I see. Yes, it's probably reasonable & best to move it out of the comprehension & cache it.

timj commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1762,6 +1763,105 @@ def getURI(self, ref: DatasetRef, predict: bool = False) -> ResourcePath:
+        missingRefs = [ref for ref in refs if ref.id not in records.keys()]
+
+        for ref in missingRefs:
+            # if this has never been written then we have to guess
```
This code has all been copied from `getURIs()`. We should not be doing that. Better to call `self.getURIs()` directly. The one reason not to do that is that we already know that `self.exists(ref)` is going to fail and so it would be more efficient to not call that at all.

The quick fix is to delete this code and live with doing an extra call to exists() that we know is going to fail.

The efficient thing to do is to factor out the URI prediction code so that getURIs can call it and getManyURIs can call it with getManyURIs calling it without having to do the self.exists check.

timj commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1762,6 +1763,105 @@ def getURI(self, ref: DatasetRef, predict: bool = False) -> ResourcePath:
+
+        for ref in missingRefs:
+            # if this has never been written then we have to guess
+            if not self.exists(ref) and not predict:
```
We know by definition that `self.exists(ref)` is False (because it will query the datastore records first and find that there is no entry) so this check tells us nothing. We can raise immediately `if missingRefs and not predict`.

The exists() call in getURIs is checking that the datastore knows about the file and that the file is there. getManyURIs is not checking that the file is really there, it's assuming it is if datastore thinks it is. We probably should document that distinction.

timj commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1762,6 +1763,105 @@ def getURI(self, ref: DatasetRef, predict: bool = False) -> ResourcePath:
+
+            fileLocations = [(i.file_location(self.locationFactory), i) for i in storedFileInfos]
+
+            guessing = False
```
Isn't this all duplicating the logic from the second half of getURIs? We can't be duplicating all this. Can you factor getURIs into a prediction part and "must exist" part so they can both share the same logic? The first new method takes a DatasetRef that is known not to exist, the second can take a StoredFileInfo.

timj commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1762,6 +1763,105 @@ def getURI(self, ref: DatasetRef, predict: bool = False) -> ResourcePath:
+
+            uris[ref] = DatasetRefURIs(primary, components)
+
+        for ref in refs:
```
This loop is going over all the input refs, including the ones we already dealt with. Is it worth partitioning `refs` at the start into ones that were found and ones that weren't?

timj reviewed (changes requested):
I've not looked at the test code. My main comment is that there is too much code duplication of getURIs. getURIs needs to be split into two helper methods that getManyURIs can use. The logic in getURIs is to convoluted to clone it.

n8pease committed:
change to DatasetRefURIs

n8pease committed:
add getManyURIs

n8pease committed:
satisfy black

n8pease commented:
Ok, I think I've addressed the code review comments. I'm still working on getting the last of the CI steps to pass, I guess the unit test is failing using python 3.8(?), so I'll have to figure out how to run that version of python on devl-01 (any suggestions?). And then I can look at the codecov report and try to cover missing lines.

timj commented:
The unit test is failing using python 3.9 and seemingly that causes the rest of the matrix to be killed early. The python 3.9 failure doesn't seem to be caused by python 3.9 vs 3.10 but you can edit the action work flow to leave only 3.10 in the matrix to check that it does pass with 3.10.

n8pease commented:
@timj tomorrow is my last planned full day of work. I've been working on test coverage, but it's slow going. I can spend some more time on it tomorrow. But, this might be ready for re-review, or if you want we can discuss how to hand it off? And, I'll plan to dial into the MW stand up on Thursday.

timj commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -707,10 +757,60 @@ def transfer_from(
+            If the datastore does not know about a dataset, should it
+            return a predicted URI or not?
+        allow_missing : `bool`
+            If `False`, and `predct` is `False`, will raise if a `DatasetRef`
```
Typo `predct`

timj commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -501,9 +502,37 @@ def _finishIngest(
+                    break
+
+        if missing_refs and not allow_missing:
+            raise FileNotFoundError("Dataset(s) {} not in this datastore.".format(missing_refs))
```
You can use an f-string here if you like.

timj commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1762,6 +1710,157 @@ def getURI(self, ref: DatasetRef, predict: bool = False) -> ResourcePath:
+        ref: DatasetRef,
+        allow_missing: bool,
+    ) -> DatasetRefURIs:
+        """Predict a dataset ref's URIs.
```
If you don't want to be possessive you can say something like

> Predict the URIs of a `DatasetRef`

timj commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1762,6 +1710,157 @@ def getURI(self, ref: DatasetRef, predict: bool = False) -> ResourcePath:
+        ref : `DatasetRef`
+            Reference to the required Dataset.
+        allow_missing: `bool`
+            If `False` and a ref is not able to be predicted then throw.
```
then raise? rather than throw?

timj commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1762,6 +1710,157 @@ def getURI(self, ref: DatasetRef, predict: bool = False) -> ResourcePath:
+        ----------
+        ref : `DatasetRef`
+            Reference to the required Dataset.
+        allow_missing: `bool`
```
Were you encountering problems with prediction that required this parameter? If someone is wanting a bunch of predicted URIs and then encounters a `DatasetTypeNotSupportedError` that would be surprising. Maybe this scenario is so rare we should at least also log (maybe DEBUG for now).

n8pease replied:
I tried taking it out all unit tests, Jenkins, and the demo query in the notebook work fine without it, so I'm assuming I was confused about something when I added it here. I'll leave it out.

timj commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1762,6 +1710,157 @@ def getURI(self, ref: DatasetRef, predict: bool = False) -> ResourcePath:
+        for ref in existing_refs:
+            if ref.id is None:
+                raise RuntimeError("Unexpected DatasetRef.id == None for existing refs.")
+            file_infos = records[ref.id]
```
If you use `ref.getCheckedId()` here you can remove the mypy `if ref.id is None` check above because `getCheckedId` will ensure you have a valid dataset ref.

timj commented on tests/test_cliCmdQueryDatasets.py:
```diff
@@ -44,141 +143,39 @@ def _queryDatasets(repo, glob=(), collections=(), where="", find_first=False, sh
     def testNoShowURI(self):
         """Test for expected output without show_uri (default is False)."""
-        tables = self._queryDatasets(repo=self.root)
+        _ = MetricTestRepo(self.repoDir, configFile=self.configFile)
```
Does something complain if the `_ =` is missing?

n8pease replied:
If I make it a named variable then at least my editor complains that the var is never used. I guess I could try not assigning the instance to any var, though I would think that would look like an error to a linter as well, and would probably merit a comment for humans. I was thinking that assigning to `_` is the most expressive way to say "yes, I'm not using this"

timj commented on tests/test_datastore.py:
```diff
@@ -57,6 +60,57 @@
+    """A dict-like class that can be used for a DataId dict that is hashable.
+
+    By default the class is immutable ("frozen"). The `frozen`
+    attribue can be set to `False` to change values (but note that
```
typo: attribue

timj reviewed (approved):
This is great. Thanks.

n8pease committed:
try removing allow_missing from URI prediction
